### PR TITLE
feat: Integrate @commitlint for Enhanced Commit Message Generation and Configuration Support

### PR DESCRIPTION
**Pull request by sebastienfi**
# Description
This pull request integrates the @commitlint module to enhance commit message generation through the adoption of locally defined rules, and to optimize configuration management. This enhancement not only supports switching between different prompt modules but also provides developers with detailed control over commit message structures through the application of commitlint rules.

https://github.com/di-sukharev/opencommit/issues/60
https://github.com/di-sukharev/opencommit/issues/61

## Adaptive Consistency with Commitlint Rules Through a Two-Step Transformation
One of the prime features of this integration is achieving consistency in commit messages according to locally defined dynamic commitlint rules. The process is refined into two transformation steps:

1. **LLM-Compatible Rules Inference**: Initially, commitlint rules are inferred and translated into static prompts that are compatible with LLM.
2. **OpenAI for Example Commit Message Creation**: This step utilizes OpenAI to generate a sample commit message structured in JSON, using a given commit diff. This message should ideally comply with commitlint rules at this stage. 
3. **Injecting Consistency Prompt for few-shots main prompt**: The exemplary commit message generated earlier is utilized as a consistency prompt to guide the LLM in generating the commit message for subsequent commit diffs in a few-shots technique. This ensures that the messages are consistent and adhere to the locally defined commitlint rules. 

## Summary of Changes
- Instructions were added on how to switch to the @commitlint prompt module using the `OCO_PROMPT_MODULE` environment variable. Details on how integration takes place and how to view or modify the configuration have been included.
- The `@commitlint/types` package has been added to devDependencies to ensure compatibility with the conventional commit format. The `crypto` package has been added to dependencies for cryptographic functions (hash comparison).
- Refactor of the structure of diff message generation for readability and facilitated plugging in of prompt modules.
- Added support for the `OCO_PROMPT_MODULE` configuration option allowing users to switch between '@commitlint' or the default 'conventional-commit' as the prompt module.
- Added a commitlint configuration file, utility functions, and types to the commitlint module. This includes utility functions for reading, writing, and checking the existence of the configuration file, and also for cleaning up the configuration string.

## Testing
1. Open a project which contains a commitlint configuration. (See [install step](https://commitlint.js.org/#/?id=install))
2. Set the `OCO_PROMPT_MODULE` environment variable with `oco config set OCO_PROMPT_MODULE=@commitlint` ([Locally of globally](https://github.com/di-sukharev/opencommit#local-per-repo-configuration)).
3. Verify that commit messages are generated according to the specified commitlint rules.

## Example

image

File `.commitlintrc.yml`: (volontarily set weird `type-enum` to demonstrate feature)
```md
extends:
  - "@commitlint/config-conventional"
rules:
  body-leading-blank:
    - 2
    - always
  body-max-line-length:
    - 2
    - always
    - 100
  footer-leading-blank:
    - 2
    - always
  footer-max-line-length:
    - 2
    - always
    - 100
  header-max-length:
    - 2
    - always
    - 100
  subject-case:
    - 2
    - never
    - - sentence-case
      - start-case
      - pascal-case
      - upper-case
  subject-empty:
    - 2
    - never
  subject-full-stop:
    - 2
    - never
    - .
  type-case:
    - 2
    - always
    - lower-case
  type-empty:
    - 2
    - never
  type-enum:
    - 2
    - always
    - - featuring
      - fixing
      - documenting
      - styling
      - refactoring
      - testing
      - chores
      - perf
      - build
      - ci
      - revert
  body-empty:
    - 2
    - never
  scope-empty:
    - 2
    - never
prompt:
  questions:
    type:
      description: Select the type of change that you're committing
      enum:
        featuring:
          description: A new feature
        fixing:
          description: A bug fix
        documenting:
          description: Documentation only changes
        styling:
          description: >-
            Changes that do not affect the meaning of the code (white-space,
            formatting, missing semi-colons, etc)
        refactoring:
          description: A code change that neither fixes a bug nor adds a feature
        perf:
          description: A code change that improves performance
        testing:
          description: Adding missing tests or correcting existing tests
        build:
          description: >-
            Changes that affect the build system or external dependencies
            (example scopes: gulp, broccoli, npm)
        ci:
          description: >-
            Changes to our CI configuration files and scripts (example scopes:
            Travis, Circle, BrowserStack, SauceLabs)
        chores:
          description: Other changes that don't modify src or test files
        revert:
          description: Reverts a previous commit
```

Generated prompts at config stage. File `.opencommit-commitlint`:
```json
{
  "hash": "f7473338e1a458cf3c33801f61943d31beb9532377b42e490a6ea6c4bbeb900e",
  "prompts": [
    "There should always be a blank line at the beginning of the body.",
    "Each line of the body should always have 100 characters or less.",
    "There should always be a blank line at the beginning of the footer.",
    "Each line of the footer should always have 100 characters or less.",
    "The header should always have 100 characters or less.",
    "The subject should never be in one of the following case: \n  - sentence-case\n  - start-case\n  - pascal-case\n  - upper-case.",
    "The subject should never be empty.",
    "The subject should never end with '.'.",
    "The type should always be in lower-case case.",
    "The type should never be empty.",
    "The type should always be one of the following values: \n  - featuring (A new feature)\n  - fixing (A bug fix)\n  - documenting (Documentation only changes)\n  - styling (Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc))\n  - refactoring (A code change that neither fixes a bug nor adds a feature)\n  - testing (Adding missing tests or correcting existing tests)\n  - chores (Other changes that don't modify src or test files)\n  - perf (A code change that improves performance)\n  - build (Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm))\n  - ci (Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs))\n  - revert (Reverts a previous commit).",
    "The body should never be empty.",
    "The scope should never be empty."
  ],
  "consistency": {
    "english": {
      "localLanguage": "english",
      "commitFix": "fixing(server): change port variable to uppercase",
      "commitFeat": "featuring(server): allow server to listen on environment variable specified port",
      "commitDescription": "Change port variable to uppercase (#123)\nAllow server to listen on environment variable specified port (#124)"
    }
  }
}
```

On second run, configuration already exists:
image

Please review and provide feedback

### CODE REVIEW

**Comment by sebastienfi** [comment]
Removed issue number from prompt.

File: src/generateCommitMessageFromGitDiff.ts [hunk 1]
```diff
@@ -1,228 +1,193 @@
...
+  chatContextAsCompletionRequest.push({
+    role: ChatCompletionRequestMessageRoleEnum.User,
+    content: diff
+  });
 
-    return chatContextAsCompletionRequest;
+  // fs.writeFile(
+  //   `.commitlint-debug`,
+  //   JSON.stringify(chatContextAsCompletionRequest, null, 2)
+  // );
```
Comment (BlakeEddie) [inline src/generateCommitMessageFromGitDiff.ts:28]:
```suggestion
```

**Review by BlakeEddie — COMMENTED** [review]
looks pretty good

suggest removing the debug comments if they are no longer needed
using guard clauses instead of if, elseif, else may make reading a little easier but not a big issue

**Comment by di-sukharev** [comment]
@sebastienfi sorry to be sooo late, was [winning top 1](https://dev.to/devteam/github-dev-2023-hackathon-winners-announced-236o) in github hackathon with OpenCommit :) 

that's a solid job, thank you <3 let me take some time to review it, i should have time on the weekend

**Comment by sebastienfi** [comment]
Hi @di-sukharev I have some time over the week, would you like to meet in visio and make a review together?
Reach me by email at [this address](https://privatebin.net/?f568a69b7719de5e#CsnSpQDQDvhjshSEafyPLMxa9CPRnjaHvn5Hkc4rMuJk)

**Comment by di-sukharev** [comment]
Hi 👋 

I was really busy for this 4 weeks, I should have time tomorrow finally to review and merge all the PRs hanging :)

**Comment by di-sukharev** [comment]
this is extremely great job done. thanks you @sebastienfi, let's call tomorrow and review it all together? I have 20 minutes tmr from 13:00 UTC to 14:00 UTC, is it ok?

in the meantime, could you please fix 3 conflicts left?

**Comment by sebastienfi** [comment]
@di-sukharev should this be merger now?

**Comment by di-sukharev** [comment]
@sebastienfi omg, sorry.. im so busy, i will merge this next week!

**Comment by di-sukharev** [comment]
@sebastienfi could you change the base to `dev` branch? we dont merge to master directly

**Comment by di-sukharev** [comment]
@sebastienfi hi man, this is an amazing contribution! thank you so much for your ideas and refactoring!

i've changed the base from master to dev, gonna manage this beautiful peace today. have you tested it locally?

**Review by di-sukharev — APPROVED** [review]

**Review by di-sukharev — APPROVED** [review]

**Comment by sebastienfi** [comment]
Sorry for long silence, been AFK. All good @di-sukharev ?

**Pull request merged**